instance/models/mixins/rabbitmq.py

```diff
@@ -157,13 +156,27 @@ def provision_rabbitmq(self):
         self.rabbitmq_provisioned = True
         self.save()
 
-    def deprovision_rabbitmq(self):
+    def deprovision_rabbitmq(self, ignore_errors=False):
         """
         Deletes the RabbitMQ vhost and users.
         """
+        self.logger.info('Deprovisioning RabbitMQ started.')
         if self.rabbitmq_provisioned:
-            self._rabbitmq_request('delete', 'vhosts', self.rabbitmq_vhost)
+            self.logger.info('Deleting rabbitmq vhost: %s', self.rabbitmq_vhost)
+            try:
+                self._rabbitmq_request('delete', 'vhosts', self.rabbitmq_vhost)
+            except RabbitMQAPIError as exc:
+                self.logger.exception('Cannot delete rabbitmq vhost: %s. %s', self.rabbitmq_vhost, exc)
+                if not ignore_errors:
+                    raise
             for user in [self.rabbitmq_consumer_user, self.rabbitmq_provider_user]:
-                self._rabbitmq_request('delete', 'users', user.username)
+                self.logger.info('Deleting rabbitmq user: %s', user.username)
+                try:
+                    self._rabbitmq_request('delete', 'users', user.username)
+                except RabbitMQAPIError as exc:
+                    self.logger.exception('Cannot delete rabbitmq user: %s. %s', user.username, exc)
+                    if not ignore_errors:
+                        raise
```

Comment:
Could we move it to `_rabbitmq_request` to avoid code repetition?

On the second thought, we could also simplify the whole exception handling by adding decorator for exception logging and optionally reraising the exception. Please see [this link](https://www.blog.pythonlibrary.org/2016/06/09/python-how-to-create-an-exception-logging-decorator/) for the extended idea. It would greatly increase the code readability, because we won't need to dig through the error handling here.

Reply:
I think it would indeed increase readability but I'm not sure about what it would do to debugging.
If there's anything other error that needs to be followed through the stacktrace will, unnecessarily, have another line where the actual exception was logged. And that extra set of lines will probably not even live in the same module.
If we're ok with that then the decorator should probably live in a more general `utils.py` or `decorators.py` module and we could use it in other methods that catch specific exceptions.
The decorator should catch specific exceptions so a `exc_class` argument should be there as well as the message we wish to log.
But there's another issue here. These exceptions are not in `_rabbitmq_request` because that method is too broad for what we want to log. We want to specify either which user or vhost we were not able to delete. This would also be the issue if we move error handling into a decorator since we won't necessarily have access to the variables in the method's inner scope.